### PR TITLE
Fit screen to size 

### PR DESCRIPTION
**Pull request by ivangitduck**
I deciced to implement a feature on rrweb that fits the content within some size. 

Here, we can control suchs events like host resize, and reacting to this easier. 

There is still work to do related to mourse tracking if the content is resized, but scaled the content perfectly

### CODE REVIEW

File: src/replay/index.ts [hunk 1]
```diff
@@ -117,6 +117,9 @@ export class Replayer {
 
   private newDocumentQueue: addedNodeMutation[] = [];
 
+  private offsetFactorFit = 1
+  private offsetMarginFit: number | null = null
```
Comment (ivangitduck) [inline src/replay/index.ts]: I am planning to use this vars for mouse tracking scale

File: src/replay/index.ts [hunk 2]
```diff
@@ -144,11 +147,13 @@ export class Replayer {
     if (!this.config.logConfig.replayLogger)
       this.config.logConfig.replayLogger = this.getConsoleLogger();
 
+    this.diggestAutoFit = this.diggestAutoFit.bind(this)
     this.handleResize = this.handleResize.bind(this);
     this.getCastFn = this.getCastFn.bind(this);
     this.emitter.on(ReplayerEvents.Resize, this.handleResize as Handler);
 
     this.setupDom();
+    this.setUpAutoFit.bind(this)()
```
Comment (lepsch) [inline src/replay/index.ts]: No need to bind them as they aren't used as callback anywhere
```suggestion
    this.handleResize = this.handleResize.bind(this);
    this.getCastFn = this.getCastFn.bind(this);
    this.emitter.on(ReplayerEvents.Resize, this.handleResize as Handler);

    this.setupDom();
    this.setUpAutoFit()
```

File: src/replay/index.ts [hunk 3]
```diff
@@ -420,6 +425,52 @@ export class Replayer {
     }
   }
 
+  private setUpAutoFit() {
+    var observer = new MutationObserver((mutations) => {
+      mutations.forEach((mutation) => {
+        if (mutation.type == "attributes") {
+          console.log("test attributes changed", mutation)
```
Comment (lepsch) [inline src/replay/index.ts]: Debug code
```suggestion
```

File: src/replay/index.ts [hunk 4]
```diff
@@ -420,6 +425,52 @@ export class Replayer {
...
+      if ( factor > 0.99) return
+      const margin = (iframeWidthContainer - sizeWidthContainer) / 2
+      
+      this.iframe.style.transform = `scale(${factor})`
+      this.iframe.style.marginLeft = `-${margin}px`
+
+      this.iframe.style.marginTop = `-${(iframeHeightContainer - (iframeHeightContainer * factor)) / 2}px`
+      
+      this.offsetFactorFit = 1
+      this.offsetMarginFit = margin
```
Comment (lepsch) [inline src/replay/index.ts]: Those 2 props are never read

File: src/replay/index.ts [hunk 5]
```diff
@@ -420,6 +425,52 @@ export class Replayer {
...
+    this.diggestAutoFit()
+
+    return observer
+  }
+
+  private diggestAutoFit() {
+    this.offsetMarginFit = null
+    this.offsetFactorFit = 1
+
+    if(!this.wrapper.hasAttribute('data-fit') || this.wrapper.getAttribute('data-fit') === '0') {
```
Comment (lepsch) [inline src/replay/index.ts]: As I understand this, looks like the host is going to change the wrapper attribute named `data-fit` so this triggers in the replayer. I think a better approach would be using the `CustomEvent` in the host and listening to it in the guests.

On the host you're gonna call the function `addCustomEvent` ([declared here](https://github.com/rrweb-io/rrweb/blob/924a4c898615d89c8c1cb240370c9e0ab7edaf73/typings/record/index.d.ts#L4)). Then, you can listen to custom events with `replayer.on(ReplayerEvents.CustomEvent, <callback>)` [like this](https://github.com/gitduckhq/gitduck/blob/3bf5d775f57cd67ac625a1c8a4574ed13d68b355/teletype-dom/lib/GuestDomBinding.ts#L46).

Comment (lepsch) [reply src/replay/index.ts]: Also, using `CustomEvent` there is no need to change the sources here

File: src/replay/index.ts [hunk 6]
```diff
@@ -144,6 +148,7 @@ export class Replayer {
     if (!this.config.logConfig.replayLogger)
       this.config.logConfig.replayLogger = this.getConsoleLogger();
 
+    this.diggestAutoFit = this.diggestAutoFit.bind(this)
```
Comment (lepsch) [inline src/replay/index.ts]: This function is not used as a callback so no need to bind it
```suggestion
```

File: src/replay/index.ts [hunk 7]
```diff
@@ -420,6 +425,51 @@ export class Replayer {
...
+    const hegihtDiffOnFit = iframeHeightContainer - (iframeHeightContainer * factor)
+
+    this.iframe.style.transform = `scale(${factor})`
+    this.iframe.style.marginLeft = `-${margin}px`
+
+    this.iframe.style.marginTop = `-${hegihtDiffOnFit / 2}px`
+    
+    this.offsetFactorFit = factor
+    this.offsetMarginFit = hegihtDiffOnFit / 2
+
```
Comment (lepsch) [inline src/replay/index.ts]: Typo
```suggestion
    const heightDiffOnFit = iframeHeightContainer - (iframeHeightContainer * factor)

    this.iframe.style.transform = `scale(${factor})`
    this.iframe.style.marginLeft = `-${margin}px`

    this.iframe.style.marginTop = `-${heightDiffOnFit / 2}px`
    
    this.offsetFactorFit = factor
    this.offsetMarginFit = heightDiffOnFit / 2
```

File: src/replay/index.ts [hunk 8]
```diff
@@ -420,6 +425,51 @@ export class Replayer {
...
+  public autoFitToWidth(size: number) {
+    this.offsetMarginFit = 0
+    this.offsetFactorFit = 1
+    this.currentFitSize = size
+
+    const iframeWidthContainer = parseFloat(this.iframe.width)
+    const iframeHeightContainer = parseFloat(this.iframe.height)
+    const factor = size / iframeWidthContainer
+
+    if ( factor > 0.99) {
```
Comment (lepsch) [inline src/replay/index.ts]: NIT
```suggestion
    if (factor > 0.99) {
```

File: src/replay/index.ts [hunk 9]
```diff
@@ -420,6 +425,51 @@ export class Replayer {
...
+
+    this.iframe.style.marginTop = `-${hegihtDiffOnFit / 2}px`
+    
+    this.offsetFactorFit = factor
+    this.offsetMarginFit = hegihtDiffOnFit / 2
+    
+  }
+
+  private diggestAutoFit() {
+    if(this.currentFitSize !== null) {
```
Comment (lepsch) [inline src/replay/index.ts]: NIT
```suggestion
    if (this.currentFitSize !== null) {
```

File: src/replay/index.ts [hunk 10]
```diff
@@ -420,6 +425,51 @@ export class Replayer {
     }
   }
 
+  public originalFit() {
```
Comment (lepsch) [inline src/replay/index.ts]: I'd propose renaming this to something with a verb on it so is resembles an action when you call it. The way it is looks like a property.
```suggestion
  public resetAutoFit() {
```

**Review by lepsch — APPROVED** [review]
LGTM. Some comments left though

**Pull request merged**